alibi_detect/cd/pytorch/preprocess.py

```diff
@@ -7,6 +7,63 @@
                                                    predict_batch_transformer)
 
 
+class _Encoder(nn.Module):
+    def __init__(
+            self,
+            input_layer: Union[nn.Module, int],
+            mlp: Optional[nn.Module] = None,
+            input_dim: Optional[int] = None,
+            enc_dim: Optional[int] = None,
+            step_dim: Optional[int] = None,
+    ) -> None:
+        super().__init__()
+        self.input_layer = input_layer
+        if isinstance(mlp, nn.Module):
+            self.mlp = mlp
+        elif isinstance(enc_dim, int) and isinstance(step_dim, int):
+            self.mlp = nn.Sequential(
+                nn.Flatten(),
+                nn.Linear(input_dim, enc_dim + 2 * step_dim),
+                nn.ReLU(),
+                nn.Linear(enc_dim + 2 * step_dim, enc_dim + step_dim),
+                nn.ReLU(),
+                nn.Linear(enc_dim + step_dim, enc_dim)
+            )
+        else:
+            raise ValueError('Need to provide either `enc_dim` and `step_dim` or a '
+                             'nn.Module `mlp`')
+
+    def forward(self, x: Union[np.ndarray, torch.Tensor, Dict[str, torch.Tensor]]) -> torch.Tensor:
+        if isinstance(self.input_layer, nn.Module):
+            x = self.input_layer(x)
+        return self.mlp(x)
+
+
+class UAE(nn.Module):
```

Comment:
Colab notebook for prototyping:
https://gist.github.com/ascillitoe/0c967eeac59a941a17f4c6c258b1ef81